Introduce standard [ReflectURL] extended attribute

lukewarlow opened the pull request:
#### 2f61732dd068c1a4174963be7f01c3e3da526702
<pre>
Introduce standard [ReflectURL] extended attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=296714">https://bugs.webkit.org/show_bug.cgi?id=296714</a>

Reviewed by Darin Adler.

See <a href="https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#xattr-reflecturl">https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#xattr-reflecturl</a>

This updates the IDL parser to support extended attributes that take string arguments.

It also replaces usages of [Reflect, URL] with the new [ReflectURL] to match the updated spec.

* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(ContentAttributeName):
(UnquoteStringLiteral):
(GetterExpression):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateAttributeGetterBodyDefinition):
* Source/WebCore/bindings/scripts/IDLAttributes.json:
* Source/WebCore/bindings/scripts/IDLParser.pm:
(parseExtendedAttributeRest2):
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
(WebCore::JSTestObjDOMConstructor::construct):
(WebCore::JSTestObjPrototype::finishCreation):
(WebCore::JSC_DEFINE_CUSTOM_SETTER):
(WebCore::jsTestObj_reflectedSetterURLAttrGetter): Deleted.
(WebCore::setJSTestObj_reflectedSetterURLAttrSetter): Deleted.
(WebCore::jsTestObj_reflectedSetterUSVURLAttrGetter): Deleted.
(WebCore::setJSTestObj_reflectedSetterUSVURLAttrSetter): Deleted.
(WebCore::jsTestObj_reflectedSetterCustomURLAttrGetter): Deleted.
(WebCore::setJSTestObj_reflectedSetterCustomURLAttrSetter): Deleted.
* Source/WebCore/bindings/scripts/test/TestObj.idl:
* Source/WebCore/html/HTMLEmbedElement.idl:
* Source/WebCore/html/HTMLFrameElement.idl:
* Source/WebCore/html/HTMLHyperlinkElementUtils.idl:
* Source/WebCore/html/HTMLIFrameElement.idl:
* Source/WebCore/html/HTMLImageElement.idl:
* Source/WebCore/html/HTMLInputElement.idl:
* Source/WebCore/html/HTMLLinkElement.idl:
* Source/WebCore/html/HTMLMediaElement.idl:
* Source/WebCore/html/HTMLModElement.idl:
* Source/WebCore/html/HTMLObjectElement.idl:
* Source/WebCore/html/HTMLQuoteElement.idl:
* Source/WebCore/html/HTMLSourceElement.idl:
* Source/WebCore/html/HTMLTrackElement.idl:
* Source/WebCore/html/HTMLVideoElement.idl:

Canonical link: <a href="https://commits.webkit.org/298101@main">https://commits.webkit.org/298101@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6440da9cf7794b2a7f4b87f061e4bf69928fba8f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114232 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33979 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24441 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120398 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64968 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116121 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34608 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42541 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86811 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117180 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27558 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102597 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67200 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26740 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20723 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64092 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96933 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20840 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123615 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41251 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30759 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95641 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41628 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98799 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95424 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24324 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40567 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18396 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41130 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46641 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40736 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44042 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42487 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebCore/bindings/scripts/test/TestObj.idl:
```diff
@@ -124,12 +124,9 @@ enum TestConfidence { "high", "kinda-low" };
     [ReflectSetter] attribute boolean reflectedSetterBooleanAttr;
     [ReflectSetter] attribute Element reflectedSetterElementAttr;
     [ReflectSetter] attribute FrozenArray<Element> reflectedSetterElementsArrayAttr;
-    [ReflectSetter, URL] attribute DOMString reflectedSetterURLAttr;
```
Per spec you can't do ReflectSetter and ReflectURL together so this just removes the related test cases (this is never used in practice anyway)

darinadler replied:
The specification includes `ReflectSetter`? I thought I invented it recently and didn't expect it to be specified!

If they can't be used together should we check and emit an error if someone tries?

weinig commented on Source/WebCore/bindings/scripts/test/TestObj.idl:
```diff
@@ -109,12 +109,12 @@ enum TestConfidence { "high", "kinda-low" };
     [Reflect=customContentIntegralAttr] attribute long reflectedCustomIntegralAttr;
     [Reflect=customContentBooleanAttr] attribute boolean reflectedCustomBooleanAttr;
-    [Reflect=customContentURLAttr, URL] attribute DOMString reflectedCustomURLAttr;
+    [ReflectURL="customContentURLAttr"] attribute DOMString reflectedCustomURLAttr;
```
Is a quoted string part of the spec? Usually these are bare identifiers.

lukewarlow replied:
Yes the spec uses real strings as a new extended attribute grammar.

lukewarlow replied:
Though interestingly WebKit's system doesn't seem to be using the raw attribute names it's using a codified enum value so it might not be possible to fully match the spec without a larger refactor. The spec largely relies on just lower casing the property name as the implicit attribute name, with edge cases like Reflect="http-equiv" (I plan to update Reflect and ReflectSetter in follow-ups).

But these tests don't actually match real usage either. Because in real usage it would be something like Reflect=custom_content which we can map change to Reflect="custom-content" and replace - with _ quite easily.

lukewarlow replied:
In my follow up for Reflect I account for the "http-equiv" to http_equiv conversion turns out that's all thats needed so it's quite simple to align with spec.

weinig commented on Source/WebCore/bindings/scripts/CodeGenerator.pm:
```diff
@@ -1127,6 +1130,14 @@ sub ContentAttributeName
     return "WebCore::${namespace}::${contentAttributeName}Attr";
 }
 
+sub UnquoteStringLiteral {
```
`{` should go on the next line.

weinig commented on Source/WebCore/bindings/scripts/CodeGenerator.pm:
```diff
@@ -1113,10 +1113,13 @@ sub ContentAttributeName
     my ($generator, $implIncludes, $interfaceName, $attribute, $getterOrSetter) = @_;
 
     my $reflect = $attribute->extendedAttributes->{Reflect};
+    my $reflectUrl = $attribute->extendedAttributes->{ReflectURL};
```
Please use all caps for URL, so this would be `my $reflectURL`, etc.

darinadler commented on Source/WebCore/bindings/scripts/IDLAttributes.json:
```diff
@@ -425,6 +425,13 @@
             "contextsAllowed": ["attribute"],
             "values": ["*"]
         },
+        "ReflectURL": {
```
You should also remove `URL` from this file unless it's still used for something.

lukewarlow replied:
URL is used in a few places still, but for purposes other than reflection. So I don't think it can be removed.

darinadler commented on Source/WebCore/bindings/scripts/CodeGenerator.pm:
```diff
@@ -1127,6 +1130,15 @@ sub ContentAttributeName
+    return $s if length($s) < 2;
+    return $s if substr($s, 0, 1) ne '"' || substr($s, -1) ne '"';
+    return substr($s, 1, -1);
+}
```
Why does this allow literals without quote marks?

This doesn't handle special characters but I guess that's OK for our uses.

lukewarlow replied:
It handles unquoted strings for the case of VALUE_IS_MISSING which is just a raw value that's unquoted.

lukewarlow replied:
I've updated this so that it dies if the value is unquoted and isn't VALUE_IS_MISSING

darinadler commented on Source/WebCore/bindings/scripts/CodeGenerator.pm:
```diff
@@ -1140,7 +1152,7 @@ sub GetterExpression
 
     my $functionName;
-    if ($attribute->extendedAttributes->{"URL"}) {
+    if ($attribute->extendedAttributes->{"ReflectURL"}) {
```
```suggestion
    if ($attribute->extendedAttributes->{ReflectURL}) {
```

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5593,8 +5593,8 @@ sub GenerateAttributeGetterBodyDefinition
     # to be AtomStrings is an implementation detail.
-    if ($attribute->extendedAttributes->{"Reflect"} && $codeGenerator->IsStringType($attribute->type)) {
-       die "Using [AtomString] on attributes marked as [Reflect] is unnecessary" if $attribute->type->extendedAttributes->{AtomString};
+    if (($attribute->extendedAttributes->{"Reflect"} || $attribute->extendedAttributes->{"ReflectURL"}) && $codeGenerator->IsStringType($attribute->type)) {
```
```suggestion
    if (($attribute->extendedAttributes->{Reflect} || $attribute->extendedAttributes->{ReflectURL}) && $codeGenerator->IsStringType($attribute->type)) {
```

darinadler reviewed (approved):
I think ideally we want an error if the argument to `ReflectURL` is not quoted. We don't want a mix of quoted strings and unquoted string-like tokens in our IDL files and an error will help us not introduce such a mix.

I think we also want an error if more than one of `Reflect`, `ReflectURL`, and `ReflectSetter` are specified.

lukewarlow commented on Source/WebCore/bindings/scripts/CodeGenerator.pm:
```diff
@@ -1113,10 +1113,13 @@ sub ContentAttributeName
+    my $reflectURL = $attribute->extendedAttributes->{ReflectURL};
     my $reflectSetter = $attribute->extendedAttributes->{ReflectSetter};
+    die "Do not use both [ReflectURL] and [Reflect] on the same attribute" if $reflectURL && $reflect;
+    die "Do not use both [ReflectURL] and [ReflectSetter] on the same attribute" if $reflectURL && $reflectSetter;
```
@darinadler here are the errors for if these are combined together fwiw

lukewarlow committed:
Introduce standard [ReflectURL] extended attribute
https://bugs.webkit.org/show_bug.cgi?id=296714

Reviewed by Darin Adler.

See https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#xattr-reflecturl

This updates the IDL parser to support extended attributes that take string arguments.

It also replaces usages of [Reflect, URL] with the new [ReflectURL] to match the updated spec.

* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(ContentAttributeName):
(UnquoteStringLiteral):
(GetterExpression):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateAttributeGetterBodyDefinition):
* Source/WebCore/bindings/scripts/IDLAttributes.json:
* Source/WebCore/bindings/scripts/IDLParser.pm:
(parseExtendedAttributeRest2):
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
(WebCore::JSTestObjDOMConstructor::construct):
(WebCore::JSTestObjPrototype::finishCreation):
(WebCore::JSC_DEFINE_CUSTOM_SETTER):
(WebCore::jsTestObj_reflectedSetterURLAttrGetter): Deleted.
(WebCore::setJSTestObj_reflectedSetterURLAttrSetter): Deleted.
(WebCore::jsTestObj_reflectedSetterUSVURLAttrGetter): Deleted.
(WebCore::setJSTestObj_reflectedSetterUSVURLAttrSetter): Deleted.
(WebCore::jsTestObj_reflectedSetterCustomURLAttrGetter): Deleted.
(WebCore::setJSTestObj_reflectedSetterCustomURLAttrSetter): Deleted.
* Source/WebCore/bindings/scripts/test/TestObj.idl:
* Source/WebCore/html/HTMLEmbedElement.idl:
* Source/WebCore/html/HTMLFrameElement.idl:
* Source/WebCore/html/HTMLHyperlinkElementUtils.idl:
* Source/WebCore/html/HTMLIFrameElement.idl:
* Source/WebCore/html/HTMLImageElement.idl:
* Source/WebCore/html/HTMLInputElement.idl:
* Source/WebCore/html/HTMLLinkElement.idl:
* Source/WebCore/html/HTMLMediaElement.idl:
* Source/WebCore/html/HTMLModElement.idl:
* Source/WebCore/html/HTMLObjectElement.idl:
* Source/WebCore/html/HTMLQuoteElement.idl:
* Source/WebCore/html/HTMLSourceElement.idl:
* Source/WebCore/html/HTMLTrackElement.idl:
* Source/WebCore/html/HTMLVideoElement.idl:

Canonical link: https://commits.webkit.org/298101@main
